rmf_battery/src/rmf_battery/agv/SimpleMotionPowerSink.cpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <rmf_battery/agv/SimpleMotionPowerSink.hpp>
+
+#include <rmf_traffic/Motion.hpp>
+#include <rmf_traffic/Time.hpp>
+
+#include <vector>
+#include <cmath>
+
+namespace rmf_battery {
+namespace agv {
+
+class SimpleMotionPowerSink::Implementation
+{
+public:
+  BatterySystem battery_system;
+  MechanicalSystem mechanical_system;
+};
+
+//==============================================================================
+SimpleMotionPowerSink::SimpleMotionPowerSink(
+  BatterySystem& battery_system,
+  MechanicalSystem& mechanical_system)
+: _pimpl(rmf_utils::make_impl<Implementation>(
+      Implementation{battery_system, mechanical_system}))
+{
+  // Do nothing
+}
+
+//==============================================================================
+const BatterySystem& SimpleMotionPowerSink::battery_system() const
+{
+  return _pimpl->battery_system;
+}
+
+//==============================================================================
+const MechanicalSystem& SimpleMotionPowerSink::mechanical_system() const
+{
+  return _pimpl->mechanical_system;
+}
+
+//==============================================================================
+namespace {
+
+double compute_friction_energy(
+  const double f,
+  const double m,
+  const double v,
+  const double dt)
+{
+  const double g = 9.81; // ms-1
+  return f * m * g * v * dt;
```

Comment:
The energy loss computed here is largely due to the resistance in the bearings of the wheels which prevents the robot from coasting indefinitely.

The rolling resistance frictional force can be expressed in a [similar manner as given by the sliding friction model](https://en.wikipedia.org/wiki/Rolling_resistance#Rolling_resistance_coefficient). Here, we first compute the frictional force `Fr = f*N = f*m*g` and then compute the work done over distance `v*dt`. The frictional energy loss due to rotating on the spot is not included as it will be negligible compared to the inertial energy loss. 

Another representation of the rolling resistance model is `Fr = b.N/r` where `r` is the radius of the wheel. Here `b` is another coefficient of friction per meter. Ultimately this can be reduced the model used above. 

Please let me know what you think.

Reply:
I see, thanks for the reference!

From the description in the documentation, I would have expected that the `friction_coefficient` was referring to the "rolling resistance coefficient with dimension of length" or `b` in the link, so I was confused about this formula not having any radius.

After thinking about it more, I can understand how it would make sense to normalize the friction coefficient to exclude the mass, if we want to allow the robots to change their mass when payloads are added or removed.

My recommendation at this point would be to keep this formula the way it is and just add that link to the documentation for `friction_coefficient` and explicitly state that it refers to `Crr`, the dimensionless rolling resistance coefficient.

Reply:
I've updated the documentation to reference the dimensionless coefficient. 
https://github.com/osrf/rmf_core/pull/217/commits/49fdf4469866a1ec13eedc2f295e0cb0fbcfd760